src/EditorFeatures/Core/InheritanceMargin/AbstractInheritanceMarginService.cs

```diff
@@ -18,64 +18,69 @@ namespace Microsoft.CodeAnalysis.InheritanceMargin
     internal abstract partial class AbstractInheritanceMarginService : IInheritanceMarginService
     {
         /// <summary>
-        /// Given the root and the searching span,
+        /// Given the syntax nodes to search,
         /// get all the method, event, property and type declaration syntax nodes.
         /// </summary>
-        protected abstract ImmutableArray<SyntaxNode> GetMembers(SyntaxNode root, TextSpan spanToSearch);
+        protected abstract ImmutableArray<SyntaxNode> GetMembers(ImmutableArray<SyntaxNode> nodesToSearch);
 
         /// <summary>
-        /// Get the line number for the identifier of declaration node is declared.
+        /// Get the identifier of declaration node is declared.
         /// </summary>
-        protected abstract int GetIdentifierLineNumber(SourceText sourceText, SyntaxNode declarationNode);
+        protected abstract SyntaxToken GetIdentifier(SyntaxNode declarationNode);
 
-        public async ValueTask<ImmutableArray<InheritanceMemberItem>> GetInheritanceInfoAsync(
+        public async ValueTask<ImmutableArray<InheritanceMarginItem>> GetInheritanceMemberItemsAsync(
             Document document,
             TextSpan spanToSearch,
             CancellationToken cancellationToken)
         {
             var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
-            var allDeclarationNodes = GetMembers(root, spanToSearch);
+            var allDeclarationNodes = GetMembers(root.DescendantNodes(spanToSearch).ToImmutableArray());
```

Comment:
i'm on the fence about this DescendantNodes call.  it will cause us to look at a lot of nodes.  that isn't really awesome, esp as normally we would only have to find the TypeDecls, and that would be a *very* shallow and fast enumeration.

also, ToImmutableArray isn't really necessary here.  it will force us to realize a potentially very large array full of pointers to nodes we don't care about.

Reply:
removed ToImmutableArray call.

I agree this might cause memory & perf problem and here are my thoughts.

1. For this call,  the searching might be faster if when we do DFS, stop and start backtracking if it meets method/property/etc.., it would save a lot cost. ( I could do it in following PRs)
2. When I wrote this code, I was thinking we have many other command handler/ UI that requires knowing the inheritance chain info, we could cache these information per snapshot (maybe further, also per git commit?). But that's really a future work.

Reply:
> For this call, the searching might be faster if when we do DFS, stop and start backtracking if it meets method/property/etc.., it would save a lot cost. ( I could do it in following PRs)

I think it's much easier than that.  Just have each language walk the compilation unit, returning type declaration nodes.  They only need to walk into namespaces, and into typedecls (because of nested types).  they don't need to walk any further.  this is a very quick walk that terminates quickly and only returns what is needed.  then just process those type-decl nodes in order, gathering the members of interest.